src/varsource.rs

```diff
@@ -0,0 +1,38 @@
+/// Abstracts over reading the current process environment variables as a
+/// zero-cost abstraction to support threaded in-process testing.
+use std::collections::hash_map::HashMap;
+use std::env;
+use std::ffi::{OsStr, OsString};
+
+pub trait VarSource {
```

Comment:
I'm half tempted to suggest we do "typesafe environment variables" and gather all environment variable retrieval into one place, by putting a trait `RustupVars` which is implemented for any T which implements `VarSource` and then pass `&dyn RustupVars` around rather than `VarSource` -- perhaps that's a refactor for another time though.

Reply:
not sure how that looks different

Reply:
The idea would be that anywhere using `RustupVars` would call things like `.ram_limit()` and such, but the more I think about it the less I think it's needed right now since we don't tend to read the same variable from more than one place.